### PR TITLE
`merkledb` -- rewrite and test range proof invariants; fix proof generation/veriifcation bugs

### PR DESCRIPTION
**Pull request by danlaine**
## Why this should be merged

Yay code quality and testing

## How this works

* Rewrites range proof invariant for clarity
* Adds test for range proof invariants
* Fixes bug when generating/verifying a range proof whose greatest key has length 0

## How this was tested

New UT

### CODE REVIEW

File: x/merkledb/proof.go [hunk 1]
```diff
@@ -300,12 +305,12 @@ func (proof *RangeProof) Verify(
 		return err
 	}
 
-	largestkey := end
+	largestKey := end
```
Comment (danlaine) [inline x/merkledb/proof.go:308]: nit to use camelCase

File: x/merkledb/proof.go [hunk 2]
```diff
@@ -251,11 +251,16 @@ type RangeProof struct {
 	// they are also in [EndProof].
 	StartProof []ProofNode
 
-	// A proof of the greatest key in [KeyValues], or, if this proof contains
-	// no [KeyValues], just the root.
-	// Empty if the request for this range proof gave no upper bound
-	// on the range to fetch, unless this proof contains no [KeyValues]
-	// and [StartProof] is empty.
+	// If no upper range bound was given, [KeyValues] is empty,
```
Comment (danlaine) [inline x/merkledb/proof.go:254]: Rewritten to make the invariants more clear

File: x/merkledb/proof.go [hunk 3]
```diff
@@ -728,19 +752,20 @@ func valueOrHashMatches(value Maybe[[]byte], valueOrHash Maybe[[]byte]) bool {
...
+// If [insertChildrenLessThan] is Nothing, no children are < [insertChildrenLessThan].
+// If [insertChildrenGreaterThan] is Nothing, no children are > [insertChildrenGreaterThan].
 // Assumes [t.lock] is held.
 func addPathInfo(
 	t *trieView,
 	proofPath []ProofNode,
-	startPath path,
-	endPath path,
+	insertChildrenLessThan Maybe[path],
+	insertChildrenGreaterThan Maybe[path],
```
Comment (danlaine) [inline x/merkledb/proof.go:764]: Now uses a `Maybe` to properly handle the case where we're verifying a proof that contains only the root. Namely, suppose that we had a range proof for range [`nil`, [0]]` where the only key-value pair is for the root.  In `RangeProof.Verify` we set `largestPath` to the empty path and called `addPathInfo` with `startPath` and `endPath` both the empty root. This causes `hasLowerBound` and `hasUpperBound` in `addPathInfo` to be `false`, and we did not add children greater than the root (the empty path), resulting in proof verification failing.

In other words, we need to use a `Maybe` to disambiguate between the case where the largest key in the proof is empty and the case where we shouldn't add any proof node children on the "right" side.

File: x/merkledb/trieview.go [hunk 4]
```diff
@@ -439,15 +439,23 @@ func (t *trieView) GetRangeProof(
...
 		if err != nil {
 			return nil, err
 		}
+	} else if len(end) > 0 {
+		endProof, err = t.getProof(ctx, end)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if endProof != nil {
```
Comment (danlaine) [inline x/merkledb/trieview.go:458]: The old code could violate a `RangeProof.EndProof` invariant. Namely, if `KeyValues` contained only the root, we would not set `result.EndProof`.

File: x/merkledb/db.go [hunk 5]
```diff
@@ -1063,12 +1063,31 @@ func (db *merkleDB) VerifyChangeProof(
...
 	// into the trie so that we get the expected root ID (if this proof is valid).
-	if err := addPathInfo(view, proof.StartProof, smallestPath, largestPath); err != nil {
+	insertChildrenGreaterThan := Nothing[path]()
+	if len(largestPath) > 0 {
+		insertChildrenGreaterThan = Some(largestPath)
+	}
+	insertChildrenLessThan := Nothing[path]()
+	if len(smallestPath) > 0 {
+		insertChildrenLessThan = Some(smallestPath)
+	}
```
Comment (darioush) [inline x/merkledb/db.go:1076]: same super nit

**Review by darioush — APPROVED** [review]

File: x/merkledb/trieview.go [hunk 6]
```diff
@@ -439,15 +439,23 @@ func (t *trieView) GetRangeProof(
...
-	}
-
-	if len(end) > 0 {
-		endProof, err := t.getProof(ctx, end)
+		greatestKey := result.KeyValues[len(result.KeyValues)-1].Key
+		endProof, err = t.getProof(ctx, greatestKey)
 		if err != nil {
 			return nil, err
 		}
+	} else if len(end) > 0 {
```
Comment (StephenButtolph) [inline x/merkledb/trieview.go:452]: I feel like this could would be a lot easier to understand if we just passed `start, end Maybe[[]byte]` rather than doing this `len([]byte) > 0` stuff

File: x/merkledb/db.go [hunk 7]
```diff
@@ -1063,12 +1063,31 @@ func (db *merkleDB) VerifyChangeProof(
...
 
-	// For all the nodes along the edges of the proof, insert children < [start] and > [largestKey]
+	// For all the nodes along the edges of the proof, insert children
+	// < [insertChildrenLessThan] and > [insertChildrenGreaterThan]
 	// into the trie so that we get the expected root ID (if this proof is valid).
-	if err := addPathInfo(view, proof.StartProof, smallestPath, largestPath); err != nil {
+	insertChildrenLessThan := Nothing[path]()
+	if len(smallestPath) > 0 {
+		insertChildrenLessThan = Some(smallestPath)
+	}
```
[the remark this answers is not in the document]
Comment (danlaine) [reply x/merkledb/db.go:1073]: done in https://github.com/ava-labs/avalanchego/pull/1657

File: x/merkledb/trieview.go (hunk 6 above)
Comment (danlaine) [reply x/merkledb/trieview.go:452]: done in https://github.com/ava-labs/avalanchego/pull/1657

**Comment by patrick-ogrady** [comment]
Running on HyperSDK sync tests: https://github.com/ava-labs/hypersdk/pull/227

**Review by patrick-ogrady — APPROVED** [review]

**Review by StephenButtolph — APPROVED** [review]
